src/Common/Auth/OneTimeAuth.php

```diff
@@ -0,0 +1,177 @@
+<?php
+
+/**
+ * OneTimeAuth Class
+ * Service for Onetime token creation, routing and auth
+ *
+ * @package   OpenEMR
+ * @link      http://www.open-emr.org
+ * @author    Jerry Padgett <sjpadgett@gmail.com>
+ * @copyright Copyright (c) 2023 Jerry Padgett <sjpadgett@gmail.com>
+ * @license   https://github.com/openemr/openemr/blob/master/LICENSE GNU General Public License 3
+ */
+
+namespace OpenEMR\Common\Auth;
+
+use DateInterval;
+use DateTime;
+use OpenEMR\Common\Crypto\CryptoGen;
+use OpenEMR\Common\Logging\SystemLogger;
+use OpenEMR\Common\Utils\RandomGenUtils;
+
+class OneTimeAuth
+{
+    private $scope;
+    private $context;
+
+    public function __construct($context = 'portal', $scope = 'redirect')
+    {
+        // scope = portal/service tasks (reset, register). context = portal, patient etc.
+        $this->scope = $scope;
+        $this->context = $context;
+    }
+
+    /**
+     * @param array $p [
+     *                 'pid' => '', // required for most onetime auth
+     *                 'target_link' => '', // Onetime endpoint
+     *                 'redirect_link' => '', // Where to redirect the user after auth
+     *                 'enabled_datetime' => 'NOW', // Use a datetime if wish to enable for a future date.
+     *                 'expiry_interval' => 'PT15M', // Always PTxx{Sec,Min,Day} PeriodTime
+     *                 'email' => '']
+     * @return array|bool
+     * @throws \Exception
+     */
+    public function createPortalOneTime(array $p = []): array|bool
+    {
+        $redirect_token = '';
+        $date_base = ($p['enabled_datetime'] ?? null) ?: 'NOW';
+        $expiry = new DateTime($date_base);
+        $expiry->add(new DateInterval($p['expiry_interval'] ?? 'PT15M'));
+        $token_raw = RandomGenUtils::createUniqueToken(32);
+        $pin = RandomGenUtils::createUniqueToken(6);
+        $token_encrypt = (new CryptoGen())->encryptStandard($token_raw);
+        $token_database = $token_raw . $pin . bin2hex($expiry->format('U'));
+        if (!empty($p['pid']) && !empty($token_raw)) {
+            $query_parameters = [$token_database, $p['pid']];
+            sqlStatementNoLog("UPDATE `patient_access_onsite` SET `portal_onetime` = ? WHERE `pid` = ?", $query_parameters);
+            (new SystemLogger())->debug("New onetime token added in database.");
+        } else {
+            (new SystemLogger())->debug("Onetime failed missing PID or token creation failed");
```

Comment:
Since the function already throws exceptions, would it be better to just throw an InvalidArgumentException here?